edgraph/server.go

```diff
@@ -1570,6 +1648,73 @@ func parseRequest(qc *queryContext) error {
 	return validateQuery(qc.dqlRes.Query)
 }
 
+// addQueryIfUnique adds dummy queries in the request for checking whether predicate is unique in the db
+func addQueryIfUnique(qc *queryContext) {
+	ctx := context.WithValue(context.Background(), schema.IsWrite, false)
+	if len(qc.gmuList) == 0 {
+		return
+	}
+
+	qc.uniqueVars = map[string]string{}
+	count := 1
+	buildQuery := ""
+	for gmuIndex, gmu := range qc.gmuList {
+		for rdfIndex, pred := range gmu.Set {
+			predSchema, _ := schema.State().Get(ctx, x.NamespaceAttr(pred.Namespace, pred.Predicate))
+			if !predSchema.Unique {
+				continue
+			}
+			queryVar := fmt.Sprintf("__dgraph_uniquecheck_%v__", strconv.Itoa(count))
+			uniqueVarMapKey := strconv.Itoa(gmuIndex) + strconv.Itoa(rdfIndex)
```

Comment:
One suggestion you could try is fmt.Sprintf("%v_%v", gmuIndex, rdfIndex). But this will have string keys for the map. Ideal would be ingeger keys. Having uint32/uint64 keys are really efficient, and you could achieve that using a pairing function. Read more here https://en.wikipedia.org/wiki/Pairing_function

Reply:
we could also do what we do for matrixes either row wise counting or column wise counting.